worker/peergrouper/desired_test.go

```diff
@@ -172,71 +248,110 @@ func desiredPeerGroupTests(ipVersion TestIPVersion) []desiredPeerGroupTest {
+}
 
-			info, err := newPeerGroupInfo(trackerMap, test.statuses, test.members, mongoPort, network.SpaceName(""))
-			c.Assert(err, jc.ErrorIsNil)
+func (s *desiredPeerGroupSuite) doTestDesiredPeerGroup(c *gc.C, ipVersion TestIPVersion) {
 
-			newPeers, voting, err := desiredPeerGroup(info)
-			if test.expectErr != "" {
-				c.Assert(err, gc.ErrorMatches, test.expectErr)
-				c.Assert(newPeers, gc.IsNil)
-				continue
-			}
-			c.Assert(err, jc.ErrorIsNil)
-			c.Assert(info, gc.NotNil)
+	for ti, test := range desiredPeerGroupTests(ipVersion) {
+		c.Logf("\ntest %d: %s", ti, test.about)
+		trackerMap := make(map[string]*machineTracker)
+		for _, m := range test.machines {
+			c.Assert(trackerMap[m.Id()], gc.IsNil)
+			trackerMap[m.Id()] = m
+		}
 
-			members := make([]replicaset.Member, 0, len(newPeers))
-			for _, m := range newPeers {
-				members = append(members, *m)
-			}
+		info, err := newPeerGroupInfo(trackerMap, test.statuses, test.members, mongoPort, network.SpaceName(""))
+		c.Assert(err, jc.ErrorIsNil)
 
-			sort.Sort(membersById(members))
-			c.Assert(members, jc.DeepEquals, test.expectMembers)
-			if len(members) == 0 {
-				continue
-			}
-			for i, m := range test.machines {
-				vote, votePresent := voting[m.Id()]
-				c.Check(votePresent, jc.IsTrue)
-				c.Check(vote, gc.Equals, test.expectVoting[i], gc.Commentf("machine %s", m.Id()))
-			}
-			// Assure ourselves that the total number of desired votes is odd in
-			// all circumstances.
-			c.Assert(countVotes(members)%2, gc.Equals, 1)
-
-			// Make sure that when the members are set as required, that there
-			// is no further change if desiredPeerGroup is called again.
-			info, err = newPeerGroupInfo(trackerMap, test.statuses, members, mongoPort, network.SpaceName(""))
-			c.Assert(err, jc.ErrorIsNil)
-			c.Assert(info, gc.NotNil)
-
-			newPeers, voting, err = desiredPeerGroup(info)
-			c.Assert(err, gc.IsNil)
-			for i, m := range test.machines {
-				vote, votePresent := voting[m.Id()]
-				c.Check(votePresent, jc.IsTrue)
-				c.Check(vote, gc.Equals, test.expectVoting[i], gc.Commentf("machine %s", m.Id()))
-			}
-			c.Assert(err, jc.ErrorIsNil)
+		newPeers, voting, err := desiredPeerGroup(info)
+		if test.expectErr != "" {
+			c.Assert(err, gc.ErrorMatches, test.expectErr)
+			c.Assert(newPeers, gc.IsNil)
+			continue
 		}
-	})
+		c.Assert(err, jc.ErrorIsNil)
+		c.Assert(info, gc.NotNil)
+
+		members := make([]replicaset.Member, 0, len(newPeers))
+		for _, m := range newPeers {
+			members = append(members, *m)
+		}
+
+		sort.Sort(membersById(members))
+		c.Assert(membersToTestMembers(members), jc.DeepEquals, membersToTestMembers(test.expectMembers))
+		if len(members) == 0 {
+			continue
+		}
+		for _, m := range test.machines {
+			_, votePresent := voting[m.Id()]
+			c.Check(votePresent, jc.IsTrue)
+			//c.Check(vote, gc.Equals, test.expectVoting[i], gc.Commentf("machine %s", m.Id()))
+		}
+
+		// Assure ourselves that the total number of desired votes is odd in
```

Comment:
we can actually reintroduce this assertion, because our final logic *does* continue to preserve the "we must have an odd number"

Reply:
Done